File: app/src/main/java/com/jw/railstatistics/data/TicketRecord.kt
package com.jw.railstatistics.data

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import java.util.UUID

@Entity(tableName = "tickets")
data class TicketRecord(
    @PrimaryKey
    val id: String = UUID.randomUUID().toString(),
    var origin: String = "Unknown",
    var destination: String = "Unknown",
    var price: String = "£0.00",
    var ticketType: String = "N/A",
    var classType: String = "Standard",
    var toc: String? = null,
    var outboundDate: String = "Unknown",
    var outboundTime: String = "00:00",
    var returnDate: String = "",
    var returnTime: String = "00:00",
    var wasDelayed: Boolean = false,
    var delayDuration: String = "",
    var pendingCompensation: Boolean = false,
    var compensation: String = "",
    @TypeConverters(LoyaltyProgramConverter::class)
    var loyaltyProgram: LoyaltyProgram? = null,
    var railcard: String? = null,
    var coach: String? = null,
    var seat: String? = null,
    var tocRouteRestriction: String? = null,
    var returnGroupID: String? = null,
    var isReturn: Boolean = false,
    var ticketFormat: String = "Tickets"
)

data class LoyaltyProgram(
    var virginPoints: String? = null,
    var lnerCashValue: String? = null,
    var clubAvantiJourneys: String? = null
)

// Type converter for Room database to handle LoyaltyProgram serialization
class LoyaltyProgramConverter {
    @androidx.room.TypeConverter
    fun fromLoyaltyProgram(loyaltyProgram: LoyaltyProgram?): String? {
        if (loyaltyProgram == null) return null
        return "${loyaltyProgram.virginPoints}|${loyaltyProgram.lnerCashValue}|${loyaltyProgram.clubAvantiJourneys}"
    }

    @androidx.room.TypeConverter
    fun toLoyaltyProgram(value: String?): LoyaltyProgram? {
        if (value == null) return null
        val parts = value.split("|")
        return if (parts.size == 3) {
            LoyaltyProgram(
                virginPoints = if (parts[0] == "null") null else parts[0],
                lnerCashValue = if (parts[1] == "null") null else parts[1],
                clubAvantiJourneys = if (parts[2] == "null") null else parts[2]
            )
        } else null
    }
} 

File: IOS Code coversion/TicketRecord.swift
import Foundation

struct LoyaltyProgram: Codable, Equatable {
    var virginPoints: String?
    var lnerCashValue: String?
    var clubAvantiJourneys: String?
}

struct TicketRecord: Identifiable, Codable, Equatable {
    let id: UUID
    var origin: String
    var destination: String
    var price: String
    var ticketType: String
    var classType: String
    var toc: String?
    var outboundDate: String
    var outboundTime: String
    var returnDate: String
    var returnTime: String
    var wasDelayed: Bool
    var delayDuration: String
    var pendingCompensation: Bool
    var compensation: String
    var loyaltyProgram: LoyaltyProgram?
    var railcard: String?
    var coach: String?
    var seat: String?
    var tocRouteRestriction: String?
    var returnGroupID: UUID?
    var isReturn: Bool
    var ticketFormat: String

    init(
        id: UUID = UUID(),
        origin: String = "Unknown",
        destination: String = "Unknown",
        price: String = "£0.00",
        ticketType: String = "N/A",
        classType: String = "Standard",
        toc: String? = nil,
        outboundDate: String = "Unknown",
        outboundTime: String = "00:00",
        returnDate: String = "",
        returnTime: String = "00:00",
        wasDelayed: Bool = false,
        delayDuration: String = "",
        pendingCompensation: Bool = false,
        compensation: String = "",
        loyaltyProgram: LoyaltyProgram? = nil,
        railcard: String? = nil,
        coach: String? = nil,
        seat: String? = nil,
        tocRouteRestriction: String? = nil,
        returnGroupID: UUID? = nil,
        isReturn: Bool = false,
        ticketFormat: String = "Tickets"
    ) {
        self.id = id
        self.origin = origin
        self.destination = destination
        self.price = price
        self.ticketType = ticketType
        self.classType = classType
        self.toc = toc
        self.outboundDate = outboundDate
        self.outboundTime = outboundTime
        self.returnDate = returnDate
        self.returnTime = returnTime
        self.wasDelayed = wasDelayed
        self.delayDuration = delayDuration
        self.pendingCompensation = pendingCompensation
        self.compensation = compensation
        self.loyaltyProgram = loyaltyProgram
        self.railcard = railcard
        self.coach = coach
        self.seat = seat
        self.tocRouteRestriction = tocRouteRestriction
        self.returnGroupID = returnGroupID
        self.isReturn = isReturn
        self.ticketFormat = ticketFormat
    }
}
